friktion/friktion_swap_client/swap.py

```diff
@@ -0,0 +1,521 @@
+    def give_allowance(self, wallet: Wallet, token_acct_to_delegate: PublicKey, mint: PublicKey, allowance: int) -> bool:
+
+        client = Client(self.url)
+        token = Token(
+            client,
+            mint,
+            TOKEN_PROGRAM_ID,
+            wallet.payer
+        )
+
+        acct_info = token.get_account_info(token_acct_to_delegate)
+        # use a global authority for all swaps
+        # assert acct_info.owner == swap_order.counterparty
+        
+        token.approve(
+            token_acct_to_delegate,
+            DELEGATE_AUTHORITY_ADDRESS,
+            wallet.public_key,
+            allowance
+        )
+
+        return True
+    
+    # delegate should be the PDA of the swap order
+    def verify_allowance(self, wallet: Wallet, mint: PublicKey, token_account: PublicKey, with_assertions: bool) -> Tuple[int, int]:
+        client = Client(self.url)
+        token = Token(
+            client,
+            mint,
+            TOKEN_PROGRAM_ID,
+            wallet.payer
+        )
+        
+        acct_info = token.get_account_info(token_account)
+        print(acct_info.delegated_amount)
+
+        if with_assertions:
+            assert acct_info.delegate == DELEGATE_AUTHORITY_ADDRESS
+            assert acct_info.delegated_amount >= MIN_REQUIRED_ALLOWANCE
+
+        return (acct_info.delegated_amount, acct_info.amount)
+
+    """
+    Object used to interact with swap contract
+    Args:
+        config (ContractConfig): Configuration to setup the Contract
+    """
+
+    async def get_otoken_details_for_offer(self, offer: Offer):
+        if offer.swapOrderAddress is None:
+            raise Exception('can only get otoken details on an offer associated with an existin swap order')
+        print('swap order address = ', offer.swapOrderAddress)
+        swap_order = await self.get_swap_order_for_key(offer.swapOrderAddress)
+        options_contract = await self.get_options_contract_for_key(swap_order.options_contract)
+        ul_factor = await self._get_token_norm_factor(options_contract.underlying_mint)
+        quote_factor = await self._get_token_norm_factor(options_contract.quote_mint)
+        strike_price = (ul_factor / quote_factor) * options_contract.quote_amount / options_contract.underlying_amount if options_contract.is_call else (quote_factor / ul_factor) * options_contract.underlying_amount / options_contract.quote_amount
+        return {
+            'underlyingAsset': options_contract.underlying_mint,
+            'expiryTimestamp': options_contract.expiry_ts,
+            'isPut': not options_contract.is_call,
+            'strikeAsset': options_contract.quote_mint,
+            'strikePrice': strike_price,
+            'collateralAsset': options_contract.underlying_mint
+        }
+
+    async def _get_token_norm_factor(self, mint: PublicKey) -> int:
+        client = AsyncClient(self.url)
+        await client.is_connected()
+        
+        give_token = AsyncToken(
```

Comment:
is it possible to use Client/Token (as in `verify_allowance `) instead of their async versions AsyncClient/AsyncToken? Doing so it ease the integration with the sdk

Reply:
yeah possibly, I didn't know about these until I had written the SDK - do you think it's worth changing all of them given it's in a working state already?

Reply:
let's remain with that then 😃